### PR TITLE
✨ Add tex template support

### PR DESCRIPTION
**Pull request by universvm**

### CODE REVIEW

File: src/exporto0o.ts [hunk 1]
```diff
@@ -178,7 +179,8 @@ export async function exportToOo(
 
   try {
     console.log(`[${plugin.manifest.name}]: export command: ${cmd}`);
-    await exec(cmd, { cwd: variables.currentDir, env });
+    // It is necessary to **append** to the current TEXINPUTS - NOT REPLACE. TEXINPUTS contains the basic latex classes. 
+    await exec(cmd, { cwd: variables.currentDir, env: { TEXINPUTS: joinEnvPath(textemplateDir, process.env.TEXINPUTS) } });
```
Comment (mokeyish) [inline src/exporto0o.ts:183]: These changes are unnecessary(The `env` is important for https://github.com/mokeyish/obsidian-enhancing-export/pull/83). 


You can preset the `TEXINPUTS` here:

https://github.com/mokeyish/obsidian-enhancing-export/blob/054b98469bf64df76a6628570b8fd09dfe9c2936/src/settings.ts#L86

See here: https://github.com/mokeyish/obsidian-enhancing-export/pull/84#issuecomment-1594797523

Comment (mokeyish) [reply src/exporto0o.ts:183]: It should be :

```js
{
   'PATH': '/usr/local/bin:/Library/TeX/texbin:${PATH}',
   'TEXINPUTS': '${pluginDir}/textemplate:${TEXINPUTS}'
}

```

Comment (universvm) [reply src/exporto0o.ts]: Neat, this is very elegant.

Comment (universvm) [reply src/exporto0o.ts]: @mokeyish Sorry I missed that. Should be fixed now.

**Comment by universvm** [comment]
Wait before merging. I don't think your solution worked.

**Comment by universvm** [comment]
@mokeyish Your solution does not work for TEXINPUTS. I have tested it many times and even with something like

```
  env = setPlatformValue(
    env,
    {
      'PATH': '/usr/local/bin:/Library/TeX/texbin:${PATH}',
      'TEXINPUTS': joinEnvPath('${textemplateDir}', process.env.TEXINPUTS),
    },
    'darwin'
  );
```

The error is the same. The .sty file cannot be found. 

Personally I don't think it makes much sense to allow users to change TEXINPUTS. If users are smart enough to pass their own template, then they can just pass it as an extra argument.

**Comment by mokeyish** [comment]
@universvm Can you print out the cmd and env values ​​that can and cannot be used?  There is no magic here, look at the actual value to see what is the reason.

**Comment by universvm** [comment]
As mentioned: users should not be editing this and if they want to they can do so from the .zshrc or .bashrc

**Comment by mokeyish** [comment]
> As mentioned: users should not be editing this and if they want to they can do so from the .zshrc or .bashrc which is much better as pandoc recognizes it system-wide

If you didn't start Obsidian from terminal. Electron cannot get Env from `.bashrc`.

**Comment by universvm** [comment]
That's not the case. If you set TEXINPUTS in the .bashrc it gets recognised by Obsidian-Enhancing-Export. That's how I've tested your plugin before this was even a feature.

**Comment by mokeyish** [comment]
Can you print out the cmd and env values ​​that can and cannot be used? I need to know what is the reason.

Electron cannot get Env from .bashrc.
图片



I also spent a lot of time on this project. it's an unpaid work.

**Comment by universvm** [comment]
While your plugin cannot access .bashrc, pandoc can. It's much easier to do that than to set it within the setting, so that the user can access it from the terminal too. The only reason to set TEXINPUTS in the plugin is so that we can access our templates without the user needing to download them 

This is becoming so complicated that asking the user to place the templates in a folder is probably easier than dealing with the problem above. 

> I also spent a lot of time on this project. it's an unpaid work.

#73 and #86 were both PR that made me restart my work multiple times. I too am unpaid, I wanted a feature, I implemented it. The code works and the feature you have in mind is not useful for tex templates. If you want a different feature feel free to implement it. 

This is my last comment on this PR, for the sake of merging.

**this works**
```
await exec(cmd, { cwd: variables.currentDir, env: { TEXINPUTS: joinEnvPath(textemplateDir, process.env.TEXINPUTS) } }); 
```

These solutions do NOT work:

```
    await exec(cmd, { cwd: variables.currentDir, env: { TEXINPUTS: '.:${pluginDir}/textemplate:${TEXINPUTS}' } }); 
```

```
  env = setPlatformValue(
    env,
    {
      'PATH': '/usr/local/bin:/Library/TeX/texbin:${PATH}',
      'TEXINPUTS': '${pluginDir}/textemplate:${TEXINPUTS}'
    },
    'darwin'
  );
  
```

```
  env = setPlatformValue(
    env,
    {
      'PATH': '/usr/local/bin:/Library/TeX/texbin:${PATH}',
      'TEXINPUTS': joinEnvPath('${pluginDir}/textemplate:$', process.env.TEXINPUTS)
    },
    'darwin'
  );
```

It's not magic I know, but I don't know as much about Electron as you do. I assume it has to do with how the `exec` command works.

**Comment by mokeyish** [comment]
Sorry, I can't merge directly in an unexplained, hardcoded situation. Here the exec of js is functional programming, if the execution has different results, it must be caused by different parameters, so you need to print the cmd and env parameters. If you're not willing to print out the actual cmd and env, that's the end of the job.

**Comment by universvm** [comment]
Again you are misunderstanding how we use TEXINPUTS. We don't want the users to change this variable. We want to set it to `templateDir`. The solution you suggested does not work, probably because of how electron sets the environment or how your `setPlatformValue` works. 

I used the code you sent me:

## ✅ Commit [58236e6](https://github.com/mokeyish/obsidian-enhancing-export/pull/87/commits/58236e6d0d706b3e6506f82332628aa2cd346b3a) 
Screenshot 2023-06-21 at 19 12 13


## ❌ Commit [344704a](https://github.com/mokeyish/obsidian-enhancing-export/pull/87/commits/344704ab08deb43b8ce69bab27aa66ce7d11f7b8)

Screenshot 2023-06-21 at 19 07 47

Screenshot 2023-06-21 at 19 07 55

**Comment by universvm** [comment]
@mokeyish are you planning on merging [58236e6](https://github.com/mokeyish/obsidian-enhancing-export/pull/87/commits/58236e6d0d706b3e6506f82332628aa2cd346b3a) at all?

**Comment by mokeyish** [comment]
Can you delete the data.json in the PluginDir? It seems that a empty env stored in the data.json. That's why env not work.

Show me your data.json as well.

**Comment by universvm** [comment]
@mokeyish , once again, we don't want users changing this variable. It should not be configurable from the settings. This is why:

## Scenario 1

The user changes TEXINPUTS to their own folder eg:

`TEXINPUTS='documents/templates'`

**Error:** `article.cls not found`

**Reason:** The user changed TEXINPUTS and now basic latex files are not found

## Scenario 2 

The user appends TEXINPUTS to their own folder eg:

`TEXINPUTS='documents/templates':TEXINPUTS `

**Error:** Obsidian Enhancing Export templates will not work

**Reason:** TemplateDir is not indexed

Have you used `.tex` templates before? If not, please trust me on this... TemplateDir should be appended to TEXINPUTS and should not be configurable from the app. People that need to configure it are probably familiar with .bashrc which is a better and system-wide way to add templates (and it will be seen by Obsidian enhancing Export)

**Comment by mokeyish** [comment]
image

Here is the reason why Commit [344704a](https://github.com/mokeyish/obsidian-enhancing-export/pull/87/commits/344704ab08deb43b8ce69bab27aa66ce7d11f7b8) doesn't work.  But you don't seem to care about figuring out why and fixing the problem. And blindly want to merge your hardcoded code.

If you don't want to or don't have time to continue, I suggest stopping the work. 
I also don't want to waste time on this work. To be sure, I definitely wouldn't merge your hardcoded code directly.

> Can you delete the data.json in the PluginDir? It seems that a empty env stored in the data.json. That's why env not work.
> 
> Show me your data.json as well.

**Comment by universvm** [comment]
@mokeyish you don't seem to care about reading my message either. I am not saying I don't want to fix the problem, I am saying I don't want to make TEXINPUTS configurable. Can you reply to this?

**Is your intention to make TEXINPUTS configurable from settings?**


Yes or no.

**Comment by universvm** [comment]
The only way I will accept the changes is if we add templateDir to the TEXINPUTS from the code and that it won't be removed from the settings if the user adds another template folder. 

if your intention is to add it as a configurable path, then it is a mistake because the plugin won't work. In that case it makes no sense to continue.

**Comment by mokeyish** [comment]
> @mokeyish you don't seem to care about reading my message either. 

I've asked you to print cmd and env many times.. https://github.com/mokeyish/obsidian-enhancing-export/pull/87#issuecomment-1600063033 . If you print it out at the beginning, then we'll quickly pinpoint what the reason is.


Yes, I want to make all environment variables configurable in the advanced settings. This is flexible enough to accommodate other possible configuration of environment variables.

If there are any problems, they can be solved.

**Comment by mokeyish** [comment]
If you can show me your data.json , I can also see how to add the environment variable merge feature. But neither do you.

**Comment by universvm** [comment]
I already solved the problem man, I am asking you to think about this implementation carefully. 

If the user sets its own path and modifies the TEXINPUTS from the config, we lose all the templates in templatedir. This means that we should always append templateDir to TEXINPUTS. 

I 

> @mokeyish , once again, we don't want users changing this variable. It should not be configurable from the settings. This is why:
> 
> ## Scenario 1
> The user changes TEXINPUTS to their own folder eg:
> 
> `TEXINPUTS='documents/templates'`
> 
> **Error:** `article.cls not found`
> 
> **Reason:** The user changed TEXINPUTS and now basic latex files are not found
> 
> ## Scenario 2
> The user appends TEXINPUTS to their own folder eg:
> 
> `TEXINPUTS='documents/templates':TEXINPUTS `
> 
> **Error:** Obsidian Enhancing Export templates will not work
> 
> **Reason:** TemplateDir is not indexed
> 
> Have you used `.tex` templates before? If not, please trust me on this... TemplateDir should be appended to TEXINPUTS and should not be configurable from the app. People that need to configure it are probably familiar with .bashrc which is a better and system-wide way to add templates (and it will be seen by Obsidian enhancing Export)

Essentially, TEXINPUTS should _always_ index TemplateDir. If you are okay with this I can make the changes to append TemplateDir to the user defined template and push the code. 

If you do want the users to replace TemplateDir from TEXINPUTS, that's a mistake because the plugin won't work anymore. 

if you are happy with appending the user-defined template rather than replace templateDir, then I will push the fix.

**Comment by mokeyish** [comment]
I don't know what your data.json looks like right now. My idea is to add the system's initial Env value. I guess your configuration is an empty object `{}`（no TEXINPUTS configured）, in this case, I would like to make the following improvements:
- It is only ignored if the user explicitly configures `TEXINPUTS=""` or `TEXINPUTS="path/to/textemplate"`, otherwise the initial TEXINPUTS will be added to the environment variable.

**Comment by universvm** [comment]
@mokeyish see the last commit. I did what you wanted. We can squash and merge if you are okay with the changes.

**Comment by mokeyish** [comment]
> @mokeyish see the last commits. I did what you wanted. We can squash and merge if you are okay with the changes.

LGTM, We are ready to merge. Thanks for the contribution. 

The last thing to do is embed your template files like the lua scirpts. Please look at this commit https://github.com/mokeyish/obsidian-enhancing-export/commit/56c99cd36b250a28497dc2c33bc4acf1f46960c0

**Comment by universvm** [comment]
@mokeyish  Sorry I don't get it. The files are already in the folder no? What do you mean by "embed"?

**Comment by universvm** [comment]
@mokeyish could you check the last commit? Is this what you meant?

**Comment by mokeyish** [comment]
> @mokeyish could you check the last commit? Is this what you meant?

Yes, It looks good to me.

**Comment by mokeyish** [comment]
> @mokeyish Sorry I don't get it. The files are already in the folder no? What do you mean by "embed"?

Your template file also needs to be embedded in main.js, so that it will be released to the hard disk at startup.

**Comment by mokeyish** [comment]
You can delete other files in the plugin directory except main.js, manifest.json, style.css to test whether the embedded resource files are released successfully.

**Comment by universvm** [comment]
Ok I see, let me test the templates first before merging

File: src/resources.ts [hunk 2]
```diff
@@ -5,4 +5,6 @@ const embed = (dir: string, res: Record<string, { default: Uint8Array }>) =>
 export default [
   // For other file types, the Loader must be configured in the <root>/vite.config.ts.
   embed('lua', import.meta.glob<{ default: Uint8Array }>('lua/*.lua', { eager: true })),
+  embed('tex', import.meta.glob<{ default: Uint8Array }>('textemplate/*.tex', { eager: true })),
```
Comment (mokeyish) [inline src/resources.ts:8]: First argument of embed is the directory to releases. It should be `textemplate`?

Comment (universvm) [reply src/resources.ts]: Yeah I had just noticed that. The new commit should fix it.

**Comment by mokeyish** [comment]
If you test ok, I'll start merging.

**Comment by universvm** [comment]
I tested .PDF export with: None, Dissertation, and Academic Templates. All produced the correct PDF so I think everything is working.

**Comment by mokeyish** [comment]
Glad that finally we can merge this feature. Thanks you again.

**Comment by universvm** [comment]
@mokeyish thanks! Can you test PDF templates on windows by any chance? I don't have access to a windows computer unfortunately.

**Pull request merged**